File: internal/lambda/strategies/marketmaking.go
package strategies

import (
	"context"
	"fmt"
	"log"
	"strconv"
	"sync/atomic"

	"github.com/coachpo/meltica/internal/schema"
)

// MarketMaking implements a simple market making strategy that places buy and sell orders
// around the current mid-price with configurable spreads.
type MarketMaking struct {
	Lambda interface {
		Logger() *log.Logger
		GetMarketState() MarketState
		GetLastPrice() float64
		GetBidPrice() float64
		GetAskPrice() float64
		IsTradingActive() bool
		SubmitOrder(ctx context.Context, side schema.TradeSide, quantity string, price *float64) error
	}

	// Configuration
	SpreadBps     float64 // Spread in basis points (100 bps = 1%)
	OrderSize     string  // Order size as string
	MaxOpenOrders int32   // Maximum number of open orders per side

	// State
	activeBuyOrders  atomic.Int32
	activeSellOrders atomic.Int32
	lastQuotePrice   atomic.Value // float64
}

var marketMakingSubscribedEvents = []schema.CanonicalType{
	schema.CanonicalType("TRADE"),
	schema.CanonicalType("TICKER"),
	schema.CanonicalType("EXECUTION.REPORT"),
}

// SubscribedEvents returns the list of event types this strategy subscribes to.
func (s *MarketMaking) SubscribedEvents() []schema.CanonicalType {
	return append([]schema.CanonicalType(nil), marketMakingSubscribedEvents...)
}

// MarketState represents the current market snapshot.
type MarketState struct {
	LastPrice float64
	BidPrice  float64
	AskPrice  float64
	Spread    float64
	SpreadPct float64
}

// OnTrade checks if we need to refresh quotes when trades occur.
func (s *MarketMaking) OnTrade(ctx context.Context, _ *schema.Event, _ schema.TradePayload, price float64) {
	if !s.Lambda.IsTradingActive() {
		return
	}

	market := s.Lambda.GetMarketState()
	if market.BidPrice <= 0 || market.AskPrice <= 0 {
		return
	}

	// Check if we need to requote based on price movement
	lastQuote, ok := s.lastQuotePrice.Load().(float64)
	if !ok || lastQuote == 0 {
		s.placeQuotes(ctx, market)
		return
	}

	// Requote if price moved more than half our spread
	priceMove := absFloat(price-lastQuote) / lastQuote * 10000 // in bps
	if priceMove > s.SpreadBps/2 {
		s.Lambda.Logger().Printf("[MM] Price moved %.2f bps, requoting", priceMove)
		s.placeQuotes(ctx, market)
	}
}

// OnTicker updates quotes on ticker updates.
func (s *MarketMaking) OnTicker(ctx context.Context, _ *schema.Event, _ schema.TickerPayload) {
	if !s.Lambda.IsTradingActive() {
		return
	}

	market := s.Lambda.GetMarketState()
	if market.BidPrice <= 0 || market.AskPrice <= 0 {
		return
	}

	s.placeQuotes(ctx, market)
}

// OnBookSnapshot does nothing.
func (s *MarketMaking) OnBookSnapshot(_ context.Context, _ *schema.Event, _ schema.BookSnapshotPayload) {
}

// OnOrderFilled decrements active order count.
func (s *MarketMaking) OnOrderFilled(_ context.Context, _ *schema.Event, payload schema.ExecReportPayload) {
	s.decrementActiveOrders(payload.Side)
	s.Lambda.Logger().Printf("[MM] Order filled: side=%s price=%s qty=%s",
		payload.Side, payload.AvgFillPrice, payload.FilledQuantity)
}

// OnOrderRejected decrements active order count.
func (s *MarketMaking) OnOrderRejected(_ context.Context, _ *schema.Event, payload schema.ExecReportPayload, reason string) {
	s.decrementActiveOrders(payload.Side)
	s.Lambda.Logger().Printf("[MM] Order rejected: side=%s reason=%s", payload.Side, reason)
}

// OnOrderPartialFill logs partial fills.
func (s *MarketMaking) OnOrderPartialFill(_ context.Context, _ *schema.Event, payload schema.ExecReportPayload) {
	s.Lambda.Logger().Printf("[MM] Partial fill: side=%s filled=%s remaining=%s",
		payload.Side, payload.FilledQuantity, payload.RemainingQty)
}

// OnOrderCancelled decrements active order count.
func (s *MarketMaking) OnOrderCancelled(_ context.Context, _ *schema.Event, payload schema.ExecReportPayload) {
	s.decrementActiveOrders(payload.Side)
	s.Lambda.Logger().Printf("[MM] Order cancelled: side=%s", payload.Side)
}

func (s *MarketMaking) placeQuotes(ctx context.Context, market MarketState) {
	midPrice := (market.BidPrice + market.AskPrice) / 2
	if midPrice <= 0 {
		return
	}

	// Calculate quote prices with configured spread
	spreadMultiplier := s.SpreadBps / 10000.0 // Convert bps to decimal
	buyPrice := midPrice * (1 - spreadMultiplier)
	sellPrice := midPrice * (1 + spreadMultiplier)

	// Place buy order if we have capacity
	if s.activeBuyOrders.Load() < s.MaxOpenOrders {
		if err := s.Lambda.SubmitOrder(ctx, schema.TradeSideBuy, s.OrderSize, &buyPrice); err != nil {
			s.Lambda.Logger().Printf("[MM] Failed to submit buy order: %v", err)
		} else {
			s.activeBuyOrders.Add(1)
			s.Lambda.Logger().Printf("[MM] Placed buy order: price=%.2f size=%s", buyPrice, s.OrderSize)
		}
	}

	// Place sell order if we have capacity
	if s.activeSellOrders.Load() < s.MaxOpenOrders {
		if err := s.Lambda.SubmitOrder(ctx, schema.TradeSideSell, s.OrderSize, &sellPrice); err != nil {
			s.Lambda.Logger().Printf("[MM] Failed to submit sell order: %v", err)
		} else {
			s.activeSellOrders.Add(1)
			s.Lambda.Logger().Printf("[MM] Placed sell order: price=%.2f size=%s", sellPrice, s.OrderSize)
		}
	}

	s.lastQuotePrice.Store(midPrice)
}

func (s *MarketMaking) decrementActiveOrders(side schema.TradeSide) {
	switch side {
	case schema.TradeSideBuy:
		if count := s.activeBuyOrders.Add(-1); count < 0 {
			s.activeBuyOrders.Store(0)
		}
	case schema.TradeSideSell:
		if count := s.activeSellOrders.Add(-1); count < 0 {
			s.activeSellOrders.Store(0)
		}
	}
}

func absFloat(x float64) float64 {
	if x < 0 {
		return -x
	}
	return x
}

// OnOrderAcknowledged tracks acknowledged orders (no-op for this strategy).
func (s *MarketMaking) OnOrderAcknowledged(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload) {
}

// OnOrderExpired tracks expired orders (no-op for this strategy).
func (s *MarketMaking) OnOrderExpired(_ context.Context, _ *schema.Event, _ schema.ExecReportPayload) {
}

// OnKlineSummary tracks kline data (no-op for this strategy).
func (s *MarketMaking) OnKlineSummary(_ context.Context, _ *schema.Event, _ schema.KlineSummaryPayload) {
}

// OnControlAck tracks control acknowledgments (no-op for this strategy).
func (s *MarketMaking) OnControlAck(_ context.Context, _ *schema.Event, _ schema.ControlAckPayload) {}

// OnControlResult tracks control results (no-op for this strategy).
func (s *MarketMaking) OnControlResult(_ context.Context, _ *schema.Event, _ schema.ControlResultPayload) {
}

// ParseFloat safely parses a string to float64.
func ParseFloat(s string) (float64, error) {
	if s == "" {
		return 0, fmt.Errorf("empty string")
	}
	val, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return 0, fmt.Errorf("failed to parse float: %w", err)
	}
	return val, nil
}

File: strategies/marketmaking/v1.0.0/marketmaking.js
const DEFAULT_SPREAD_BPS = 25;
const DEFAULT_ORDER_SIZE = "1";
const DEFAULT_MAX_OPEN = 2;

function parseFloatConfig(value, fallback) {
  const num = Number(value);
  if (!Number.isFinite(num)) {
    return fallback;
  }
  return num;
}

function parseIntConfig(value, fallback) {
  const num = Number(value);
  if (!Number.isFinite(num) || num < 0) {
    return fallback;
  }
  return Math.floor(num);
}

function selectProvider(runtime) {
  if (!runtime) {
    return null;
  }
  const providers = typeof runtime.providers === 'function' ? runtime.providers() : [];
  if (!Array.isArray(providers) || providers.length === 0) {
    return null;
  }
  const selector = typeof runtime.selectProvider === 'function' ? runtime.selectProvider : null;
  if (selector) {
    try {
      const choice = selector(Date.now());
      if (choice) {
        return choice;
      }
    } catch (err) {
      // fall back below
    }
  }
  return providers[0];
}

function toNumber(value) {
  if (typeof value === 'number') {
    return Number.isFinite(value) ? value : NaN;
  }
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : NaN;
}

module.exports = {
  metadata: {
    name: 'marketmaking',
    version: '1.0.0',
    displayName: 'Market Making',
    description: 'Quotes bid/ask orders around the mid price to capture spread.',
    config: [
      {
        name: 'spread_bps',
        type: 'float',
        description: 'Spread in basis points',
        default: DEFAULT_SPREAD_BPS,
        required: false,
      },
      {
        name: 'order_size',
        type: 'string',
        description: 'Quoted order size',
        default: DEFAULT_ORDER_SIZE,
        required: false,
      },
      {
        name: 'max_open_orders',
        type: 'int',
        description: 'Maximum concurrent orders per side',
        default: DEFAULT_MAX_OPEN,
        required: false,
      },
      {
        name: 'dry_run',
        type: 'bool',
        description: 'When true, strategy logs intended orders without submitting them',
        default: true,
        required: false,
      },
    ],
    events: ['Trade', 'Ticker', 'ExecReport', 'BalanceUpdate', 'RiskControl'],
  },
  create: function (env) {
    const config = env.config ?? {};
    const runtime = env.runtime ?? {};
    const log = (...parts) => env.helpers.log('[MM]', ...parts);

    const spreadBps = parseFloatConfig(config.spread_bps, DEFAULT_SPREAD_BPS);
    const orderSize = String(config.order_size ?? DEFAULT_ORDER_SIZE).trim() || DEFAULT_ORDER_SIZE;
    const maxOpen = Math.max(0, parseIntConfig(config.max_open_orders, DEFAULT_MAX_OPEN));
    const configuredDryRun = config.dry_run !== undefined ? Boolean(config.dry_run) : true;

    let activeBuy = 0;
    let activeSell = 0;
    let lastQuotePrice = 0;

    function isTradingActive() {
      if (runtime && typeof runtime.isTradingActive === 'function') {
        try {
          return Boolean(runtime.isTradingActive());
        } catch (err) {
          log('isTradingActive helper failed:', err?.message ?? err);
        }
      }
      return false;
    }

    function isDryRun() {
      if (configuredDryRun) {
        return true;
      }
      if (runtime && typeof runtime.isDryRun === 'function') {
        try {
          return Boolean(runtime.isDryRun());
        } catch (err) {
          log('isDryRun helper failed:', err?.message ?? err);
        }
      }
      return false;
    }

    function marketState() {
      if (runtime && typeof runtime.getMarketState === 'function') {
        try {
          return runtime.getMarketState();
        } catch (err) {
          log('getMarketState helper failed:', err?.message ?? err);
        }
      }
      return {
        bid: runtime && typeof runtime.getBidPrice === 'function' ? runtime.getBidPrice() : NaN,
        ask: runtime && typeof runtime.getAskPrice === 'function' ? runtime.getAskPrice() : NaN,
      };
    }

    function submitLimit(provider, side, price) {
      if (isDryRun()) {
        if (side === 'buy') {
          activeBuy += 1;
        } else {
          activeSell += 1;
        }
        log(`[DRY-RUN] Would place ${side.toUpperCase()} order on ${provider}: price=${price.toFixed(2)} size=${orderSize}`);
        return;
      }
      if (!runtime || typeof runtime.submitOrder !== 'function') {
        throw new Error('submitOrder helper unavailable');
      }
      runtime.submitOrder(provider, side, orderSize, price);
      if (side === 'buy') {
        activeBuy += 1;
      } else {
        activeSell += 1;
      }
      log(`Placed ${side.toUpperCase()} order on ${provider}: price=${price.toFixed(2)} size=${orderSize}`);
    }

    function placeQuotes(state) {
      const provider = selectProvider(runtime);
      if (!provider) {
        log('Unable to select provider');
        return;
      }
      const bid = toNumber(state?.bid);
      const ask = toNumber(state?.ask);
      if (!Number.isFinite(bid) || !Number.isFinite(ask) || bid <= 0 || ask <= 0) {
        return;
      }
      const mid = (bid + ask) / 2;
      if (!Number.isFinite(mid) || mid <= 0) {
        return;
      }

      const spreadMultiplier = spreadBps / 10000;
      const buyPrice = mid * (1 - spreadMultiplier);
      const sellPrice = mid * (1 + spreadMultiplier);

      if (activeBuy < maxOpen) {
        submitLimit(provider, 'buy', buyPrice);
      }
      if (activeSell < maxOpen) {
        submitLimit(provider, 'sell', sellPrice);
      }

      lastQuotePrice = mid;
    }

    function decrement(side) {
      if (side === 'buy') {
        activeBuy = Math.max(0, activeBuy - 1);
      } else if (side === 'sell') {
        activeSell = Math.max(0, activeSell - 1);
      }
    }

    return {
      wantsCrossProviderEvents: function () {
        return false;
      },
      onTrade: function (ctx, _evt, _payload, price) {
        if (!isTradingActive()) {
          return;
        }
        const numericPrice = toNumber(price);
        if (!Number.isFinite(numericPrice) || numericPrice <= 0) {
          return;
        }
        const state = marketState();
        if (!Number.isFinite(state?.bid) || !Number.isFinite(state?.ask)) {
          return;
        }
        if (lastQuotePrice === 0) {
          placeQuotes(state);
          return;
        }
        const priceMoveBps = Math.abs(numericPrice - lastQuotePrice) / lastQuotePrice * 10000;
        if (priceMoveBps > spreadBps / 2) {
          log(`Price moved ${priceMoveBps.toFixed(2)} bps, requoting`);
          placeQuotes(state);
        }
      },
      onTicker: function () {
        if (!isTradingActive()) {
          return;
        }
        const state = marketState();
        if (!Number.isFinite(state?.bid) || !Number.isFinite(state?.ask)) {
          return;
        }
        placeQuotes(state);
      },
      onOrderFilled: function (_ctx, _evt, payload) {
        decrement((payload?.side ?? payload?.Side ?? '').toLowerCase());
        log(
          'Order filled:',
          `side=${payload?.side ?? payload?.Side ?? ''}`,
          `price=${payload?.avgFillPrice ?? payload?.avg_fill_price ?? ''}`,
          `qty=${payload?.filledQuantity ?? payload?.filled_quantity ?? ''}`
        );
      },
      onOrderRejected: function (_ctx, _evt, payload, reason) {
        decrement((payload?.side ?? payload?.Side ?? '').toLowerCase());
        log('Order rejected:', reason ?? '');
      },
      onOrderPartialFill: function (_ctx, _evt, payload) {
        log(
          'Partial fill:',
          `side=${payload?.side ?? payload?.Side ?? ''}`,
          `filled=${payload?.filledQuantity ?? payload?.filled_quantity ?? ''}`,
          `remaining=${payload?.remainingQty ?? payload?.remaining_qty ?? ''}`
        );
      },
      onOrderCancelled: function (_ctx, _evt, payload) {
        decrement((payload?.side ?? payload?.Side ?? '').toLowerCase());
        log('Order cancelled:', payload?.side ?? payload?.Side ?? '');
      },
      onBalanceUpdate: function (_ctx, _evt, payload) {
        log(
          'Balance update:',
          `currency=${payload?.currency ?? ''}`,
          `total=${payload?.total ?? ''}`,
          `available=${payload?.available ?? ''}`
        );
      },
      onRiskControl: function (_ctx, _evt, payload) {
        activeBuy = 0;
        activeSell = 0;
        lastQuotePrice = 0;
        log(
          'Risk control notification:',
          `status=${payload?.status ?? ''}`,
          `breach=${payload?.breachType ?? payload?.breach_type ?? ''}`,
          `reason=${payload?.reason ?? ''}`
        );
      },
      onBookSnapshot: function () {},
      onOrderAcknowledged: function () {},
      onOrderExpired: function () {},
      onKlineSummary: function () {},
      onInstrumentUpdate: function () {},
    };
  },
};
